File: internal/db/merge.go
package db

import (
	"sort"

	"github.com/meteormin/wuwa-tracker/internal/types"
)

// MergeRecords는 새로 가져오거나 업로드한 가챠 기록(newRecords)과 기존 DB의 기록(dbRecords)을 병합합니다.
// 두 슬라이스는 모두 최신순(인덱스 0이 가장 최신)으로 정렬되어 있어야 합니다.
func MergeRecords(dbRecords, newRecords []types.Record) []types.Record {
	if len(dbRecords) == 0 {
		return newRecords
	}
	if len(newRecords) == 0 {
		return dbRecords
	}

	// 1. 시퀀스 매칭을 통한 오버랩 찾기
	// newRecords의 특정 인덱스 k부터 시작하는 접미사(suffix)가 dbRecords의 접두사(prefix)와 일치하는지 확인합니다.
	// 이 방식은 10연차 내에 존재하는 동일한 아이템 중복까지 순서를 보존하며 완벽하게 처리합니다.
	for k := range newRecords {
		suffixLen := len(newRecords) - k
		if suffixLen > len(dbRecords) {
			continue
		}

		match := true
		for i := range suffixLen {
			r1 := newRecords[k+i]
			r2 := dbRecords[i]
			if r1.ResourceID != r2.ResourceID || r1.Time != r2.Time {
				match = false
				break
			}
		}

		if match {
			// 오버랩 지점 찾기, newRecords[0:k]가 DB에 없는 새로운 기록들입니다.
			// 이를 기존 DB 기록의 맨 앞에 붙여 병합합니다.
			merged := make([]types.Record, 0, k+len(dbRecords))
			merged = append(merged, newRecords[0:k]...)
			merged = append(merged, dbRecords...)
			return merged
		}
	}

	// 2. 오버랩이 매칭되지 않는 경우 (6개월 이상의 공백이 있거나 서로 다른 데이터셋인 경우)
	// 시간대 비교를 통해 통째로 앞/뒤에 붙일 수 있는지 판단합니다.
	newestDBTime := dbRecords[0].Time
	oldestNewTime := newRecords[len(newRecords)-1].Time

	if oldestNewTime >= newestDBTime {
		// 새로운 기록이 기존 기록보다 전부 최신인 경우
		merged := make([]types.Record, 0, len(newRecords)+len(dbRecords))
		merged = append(merged, newRecords...)
		merged = append(merged, dbRecords...)
		return merged
	}

	newestNewTime := newRecords[0].Time
	oldestDBTime := dbRecords[len(dbRecords)-1].Time

	if newestNewTime <= oldestDBTime {
		// 새로운 기록이 기존 기록보다 전부 과거인 경우
		merged := make([]types.Record, 0, len(newRecords)+len(dbRecords))
		merged = append(merged, dbRecords...)
		merged = append(merged, newRecords...)
		return merged
	}

	// 3. 최악의 케이스: 시간대가 서로 교차하지만 시퀀스 매칭이 실패한 경우 (일부 데이터 유실 등)
	// 시간 기반 유니온 병합을 수행하여 유실 없이 합칩니다.
	return unionMerge(dbRecords, newRecords)
}

// unionMerge는 시간대별로 기록들을 묶어 중복을 최소화하면서 병합합니다.
func unionMerge(dbRecords, newRecords []types.Record) []types.Record {
	groups := make(map[string][]types.Record)
	allTimesMap := make(map[string]bool)

	for _, r := range dbRecords {
		groups[r.Time] = append(groups[r.Time], r)
		allTimesMap[r.Time] = true
	}

	for _, r := range newRecords {
		if !allTimesMap[r.Time] {
			groups[r.Time] = append(groups[r.Time], r)
			allTimesMap[r.Time] = true
		}
	}

	newTimesMap := make(map[string]bool)
	for _, r := range newRecords {
		newTimesMap[r.Time] = true
	}

	// 양쪽에 모두 존재하는 시간대의 경우 아이템 ID별 최대 빈도수를 보존하여 병합
	for t := range groups {
		if newTimesMap[t] && len(dbRecords) > 0 {
			var dbItems []types.Record
			for _, r := range dbRecords {
				if r.Time == t {
					dbItems = append(dbItems, r)
				}
			}
			var newItems []types.Record
			for _, r := range newRecords {
				if r.Time == t {
					newItems = append(newItems, r)
				}
			}

			if len(dbItems) > 0 && len(newItems) > 0 {
				dbFreq := make(map[int]int)
				dbTemplates := make(map[int]types.Record)
				for _, r := range dbItems {
					dbFreq[r.ResourceID]++
					dbTemplates[r.ResourceID] = r
				}

				newFreq := make(map[int]int)
				newTemplates := make(map[int]types.Record)
				for _, r := range newItems {
					newFreq[r.ResourceID]++
					newTemplates[r.ResourceID] = r
				}

				mergedItems := []types.Record{}
				allResourceIDs := make(map[int]bool)
				for id := range dbFreq {
					allResourceIDs[id] = true
				}
				for id := range newFreq {
					allResourceIDs[id] = true
				}

				for id := range allResourceIDs {
					count := max(newFreq[id], dbFreq[id])

					template := dbTemplates[id]
					if tRecord, ok := newTemplates[id]; ok {
						template = tRecord
					}

					for range count {
						mergedItems = append(mergedItems, template)
					}
				}
				groups[t] = mergedItems
			}
		}
	}

	merged := []types.Record{}
	for _, items := range groups {
		merged = append(merged, items...)
	}

	// 최신순으로 정렬
	sortRecords(merged)

	return merged
}

func sortRecords(s []types.Record) {
	sort.SliceStable(s, func(i, j int) bool {
		if s[i].Time != s[j].Time {
			return s[i].Time > s[j].Time
		}
		if s[i].QualityLevel != s[j].QualityLevel {
			return s[i].QualityLevel > s[j].QualityLevel
		}
		return s[i].ResourceID < s[j].ResourceID
	})
}

File: crates/wuwa-tracker-core/src/merge.rs
use crate::types::Record;
use std::collections::{BTreeMap, BTreeSet, HashMap};

pub fn merge_records(db_records: &[Record], new_records: &[Record]) -> Vec<Record> {
    if db_records.is_empty() {
        return new_records.to_vec();
    }
    if new_records.is_empty() {
        return db_records.to_vec();
    }

    for k in 0..new_records.len() {
        let suffix_len = new_records.len() - k;
        if suffix_len > db_records.len() {
            continue;
        }

        let matches = (0..suffix_len).all(|i| {
            let incoming = &new_records[k + i];
            let existing = &db_records[i];
            incoming.resource_id == existing.resource_id && incoming.time == existing.time
        });

        if matches {
            let mut merged = Vec::with_capacity(k + db_records.len());
            merged.extend_from_slice(&new_records[..k]);
            merged.extend_from_slice(db_records);
            return merged;
        }
    }

    let newest_db_time = &db_records[0].time;
    let oldest_new_time = &new_records[new_records.len() - 1].time;
    if oldest_new_time >= newest_db_time {
        let mut merged = Vec::with_capacity(new_records.len() + db_records.len());
        merged.extend_from_slice(new_records);
        merged.extend_from_slice(db_records);
        return merged;
    }

    let newest_new_time = &new_records[0].time;
    let oldest_db_time = &db_records[db_records.len() - 1].time;
    if newest_new_time <= oldest_db_time {
        let mut merged = Vec::with_capacity(new_records.len() + db_records.len());
        merged.extend_from_slice(db_records);
        merged.extend_from_slice(new_records);
        return merged;
    }

    union_merge(db_records, new_records)
}

fn union_merge(db_records: &[Record], new_records: &[Record]) -> Vec<Record> {
    let mut groups: BTreeMap<String, Vec<Record>> = BTreeMap::new();
    let mut all_times = BTreeSet::new();

    for record in db_records {
        groups
            .entry(record.time.clone())
            .or_default()
            .push(record.clone());
        all_times.insert(record.time.clone());
    }

    for record in new_records {
        if !all_times.contains(&record.time) {
            groups
                .entry(record.time.clone())
                .or_default()
                .push(record.clone());
            all_times.insert(record.time.clone());
        }
    }

    let new_times: BTreeSet<&str> = new_records
        .iter()
        .map(|record| record.time.as_str())
        .collect();
    for (time, items) in groups.iter_mut() {
        if !new_times.contains(time.as_str()) {
            continue;
        }

        let db_items: Vec<&Record> = db_records
            .iter()
            .filter(|record| record.time == *time)
            .collect();
        let incoming_items: Vec<&Record> = new_records
            .iter()
            .filter(|record| record.time == *time)
            .collect();
        if db_items.is_empty() || incoming_items.is_empty() {
            continue;
        }

        let mut db_freq = HashMap::new();
        let mut db_templates = HashMap::new();
        for record in db_items {
            *db_freq.entry(record.resource_id).or_insert(0usize) += 1;
            db_templates.insert(record.resource_id, record.clone());
        }

        let mut incoming_freq = HashMap::new();
        let mut incoming_templates = HashMap::new();
        for record in incoming_items {
            *incoming_freq.entry(record.resource_id).or_insert(0usize) += 1;
            incoming_templates.insert(record.resource_id, record.clone());
        }

        let resource_ids: BTreeSet<i32> = db_freq
            .keys()
            .chain(incoming_freq.keys())
            .copied()
            .collect();
        let mut merged_items = Vec::new();
        for resource_id in resource_ids {
            let count = db_freq
                .get(&resource_id)
                .copied()
                .unwrap_or_default()
                .max(incoming_freq.get(&resource_id).copied().unwrap_or_default());
            let template = incoming_templates
                .get(&resource_id)
                .or_else(|| db_templates.get(&resource_id))
                .expect("template exists");
            for _ in 0..count {
                merged_items.push((*template).clone());
            }
        }
        *items = merged_items;
    }

    let mut merged: Vec<Record> = groups.into_values().flatten().collect();
    merged.sort_by(|left, right| {
        right
            .time
            .cmp(&left.time)
            .then(right.quality_level.cmp(&left.quality_level))
            .then(left.resource_id.cmp(&right.resource_id))
    });
    merged
}
